### ISSUE TITLE
[feature] Use Core ML on iOS

### ISSUE THREAD

**Issue by cancan101** [tensorflow/tensorflow#10468]
Apple announced [Core ML](https://developer.apple.com/documentation/coreml) which may be useful for abstracting away the complexities of the hardware platform:

![image](https://user-images.githubusercontent.com/51059/26841383-5226ef0c-4ab7-11e7-9ce9-61849a3c0cc9.png)

Related:
* https://github.com/tensorflow/tensorflow/issues/7958 (MPS)
* https://github.com/tensorflow/tensorflow/issues/3001 (BNNS)

**Comment by ofirbb** [comment]
Actually, all that you really need is to write the implementation of the python conversion from a TF graph to a CoreML (also protobuf) model file:
[CoreML custom conversion tool](https://developer.apple.com/documentation/coreml/converting_trained_models_to_core_ml#2903105)

**Comment by michaelisard** [comment]
Thanks for opening this. We we treat this as the tracking issue, assigning to @keveman.

**Comment by johndpope** [comment]
fyi - https://github.com/hwchong/MNIST_DRAW

**Comment by songkuixi** [comment]
Does your tool support transformation from TensorFlow Inception v3 Model to CoreML Model? @johndpope

**Comment by johndpope** [comment]
The link above is a vanilla implementation of coremltools Python conversion of Kees's models (with tensorflow backend) if you search inception v3 in Keras and swap in /out code above it should work.

**Comment by johndpope** [comment]
Use this instead https://github.com/oktapodi/inception_ios_test?files=1

**Comment by gsabran** [comment]
FYI if that's helpful I've hosted the package code on github to make conversations easier: https://github.com/gsabran/coremltools

**Comment by nojvek** [comment]
Thanks @gsabran.

Is that the official pip package source or just a clone. I see apple mentions it in their docs. https://developer.apple.com/machine-learning/

I do like that Apple just didn't invent their own crazy format but built it on protobuf.

As far as I'm aware keras can load from 'json + h5'. 

```
# serialize model to JSON
model_json = model.to_json()
with open("model.json", "w") as json_file:
    json_file.write(model_json)
# serialize weights to HDF5
model.save_weights("model.h5")
print("Saved model to disk")
 
# later...
 
# load json and create model
json_file = open('model.json', 'r')
loaded_model_json = json_file.read()
json_file.close()
loaded_model = model_from_json(loaded_model_json)
# load weights into new model
loaded_model.load_weights("model.h5")
```

This should be relatively straightforward to achieve right?

**Comment by gsabran** [comment]
@nojvek it's just a clone since there's no official source on github to my knowledge

**Comment by andforce** [comment]
how convert “ckpt” to "mlmodel"?

**Comment by joelteply** [comment]
I'd suggest someone use the [caffe to tensorflow](https://github.com/ethereon/caffe-tensorflow) project as a starting point, maybe combining both, and output keras since it uses tensorflow as a backend and is compatible with coremltools.Really once you get the weights it's not horrible to do by hand, but still annoying.

**Comment by lahayeg** [comment]
Hi there. I'm novice (2 days)on tensorflow and all deep learning. I managed to train a Mobilenet model (I got a output_graph.pb and a output_labels.txt) in order to test it on iOS. But in fact the coremltools does not provide a converser. How could I manage to do it?
Option 1: tensorflow => keras => core ML
Option 2: caffe (Mobilenet seems to exist but not sure of the official compliance and I tried without to manage to install tools) => core ML

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by petewarden** [comment]
There's now an official TensorFlow to CoreML converter at https://github.com/tf-coreml/tf-coreml, so closing this as fixed.

**Comment by igor9silva** [comment]
@petewarden it doesn't seem official. This seems though -> https://github.com/apple/coremltools.

**Comment by oscarbg** [comment]
Hi @petewarden and other tensorflow devs,
new to tensoflow world so sorry I'm misinterpreting something but
just seeing Google release MLKit this week with support for Android & IOS I was curious to learn underlying tech..
seems it uses for it's custom model API tensorflow lite underneath which also supports also both Android & IOS..
on Android it uses Android NN API if avaiable to use HW accel. on phone..
problem is from quick search is that altough tensorflow lite supports IOS it doesn't uses CoreML underneath so only using CPU mode..
you say here for using CoreML there is a converter, but then you must use CoreML API and abandon Tensorflow lite API usage.. in other words "only" helps with model conversion issues..
not same inference API for both iOS & Android that is HW accel. on both OSes..
I say that because also at Google I/O also ARCore 1.2 was released and there is IOS support using ARKit underneath..
so just saying same for tensorflow lite..
of course Tensorflow API should support only reading CoreML models in case compiling for IOS..
May I open a separate issue?
thanks..
